### DISCUSSION TITLE
Why are you using "async void" in BlazoredModal.razor?

### DISCUSSION THREAD

**Discussion by nekrasov-andrii** [Blazored/Modal#422] [Q&A]
Good day Chris,
My name is Andrii, 
I use your good library "BlazoredModal" in my projects.
This is a very handy library for working with popups.
When I was review the source code, I noticed that you are using "async void" in BlazoredModal.razor file.
Why are you using "async void" in BlazoredModal.razor?
Microsoft's [guidelines](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming) say that using "async void" is bad practice.
Could you please replace "async void" with "async Task" because it's a potential threat on exceptions?
I assumed you wanted to use "await" to change the state of the component.
In methods using "async Task" you can also cause an immediate redraw of the state of the component.
I'm using the following [approach](https://wizardsofsmart.wordpress.com/2020/05/29/blazor-server-tip-sometimes-you-need-task-delay1/), which always works.
 For example:
`private async Task StateHasChangedAsync()
    {
        StateHasChanged();
        await Task.Delay(1);
    }`

**Comment by chrissainty** [comment]
Hey Nekrasov,

To be honest, I can't remember how things got into this state. I'm guessing the event handlers, originally didn't need to be async and at some point it got changed. But you're correct it, shouldn't be async void. The reason for async void is becuase of how the events are defined on the `ModalService`, they're `Action<ModalReference>` and `Action<ModalReference, ModalResult>`.

The real fix here is to change the events to be `Func<ModalReference>` and `Func<ModalReference, ModalResult>`. However, this will bring in another breaking change as it will require the `ModalService.Show` method to be async. In turn, all consumers will need to change the methods they call `ModalService.Show` in to be async, if they're not already. 

The good news is with the next release we're breaking a few things so this is going to be the best time to do it I think. I'm going to convert this to an issue so it can be worked on.

**Reply by nekrasov-andrii** [reply]
Hi Chris,
Thanks for the comprehensive answer.
I'll waiting for the next version.
Have a nice day.
